linera-core/src/chain_worker/state.rs

```diff
@@ -112,9 +129,10 @@ where
         height: BlockHeight,
         index: u32,
     ) -> Result<Option<Event>, WorkerError> {
-        self.ensure_is_active()?;
+        let mut chain = self.chain.write().await;
```

Comment:
This should be `self.chain.read()`

Reply:
How about this?
```
let chain = self.chain.read().await();
let mut chain = chain.clone_unchecked();
```

Reply:
Done. I took a deeper look, and there's no need for mutability in these operations :+1: 

I applied the suggested pattern using `clone_unchecked` in the `stage_block_execution` and `query_application` methods, but the problem is that `clone_unchecked` uses a `&mut self` receiver, so a `write` lock is needed anyway, at least to perform the clone. This limits the concurrency available, but also I'm still not sure if it's worth it, because AFAICT it only affects the client, which I think doesn't benefit much from added concurrency here. I can't affirm this with certainty, and we should measure this to decide if we should change it again later.

Reply:
Oh nooo, how badly does clone_unchecked require &mut self ? This def ruins the design that I had in mind all along.

Reply:
I tried to make it `&self` and ran into four issues:

- https://github.com/linera-io/linera-protocol/blob/bd98cdfb81bd368dd87ae382ca3e0a31954f01b0/linera-views/src/collection_view.rs#L155-L156
- https://github.com/linera-io/linera-protocol/blob/bd98cdfb81bd368dd87ae382ca3e0a31954f01b0/linera-views/src/reentrant_collection_view.rs#L174-L175
- https://github.com/linera-io/linera-protocol/blob/bd98cdfb81bd368dd87ae382ca3e0a31954f01b0/linera-views/src/key_value_store_view.rs#L264
- https://github.com/linera-io/linera-protocol/blob/bd98cdfb81bd368dd87ae382ca3e0a31954f01b0/linera-views/src/hashable_wrapper.rs#L113

In summary, we could make it `&self` if we made the method `async` and accepted the fact that we'd be locking some mutexes in order to clone their contents.

Reply:
We could do that but perhaps let's make sure that we really need it. IIRC our initial plan was for the chain actor to own a (local, mutable) `ChainStateView`.
* This local view would be used by default for everything.
* The first time a read-only view is needed for GraphQL we could clone the local view, wrap it in a `RwLock`, then remember the lock in a local variable of type `Option<Arc<Rwlock<ChainStateView>>>`.
* When the local view is about to be saved, if the rwlock is not `None`, it must be acquired in `write` mode. If the `Arc` counter is 1, then we can reset the lock to `None`. Otherwise (weird case but ok), we will clone the view and update the inner (read-only) view, immediately after saving the local view.

Am I missing something?

Reply:
Mm ok the `Arc` counter may not be `1` if the read-lock is released first, then `Arc` is decremented. But it should about to become `1`.

Reply:
Maybe I misunderstood the initial plan. I thought that with the `RwLock` we wouldn't need the `clone_unchecked` anymore. But I have now updated the PR with the implementation what you described.

There's no need to check the `Arc` counter, as long as we guarantee that:

- read and write locks are always acquired in `&mut self` methods
- the write lock is only acquired by taking the `Option<Arc<RwLock<_>>>`

As for a better construction, poisoning requires panicking, which feels weird and might be inefficient and barriers require knowing how many times the view will be shared ahead of time. One option is to use a linked-chain of notifiers. It might be cleaner, but likely requires more work to notify-on-drop, and I don't know if it would be more efficient :thinking:

Reply:
Oh right because the `Arc`-holder is hidden in `OwnedRwLockReadGuard`